# soma-cnv
Detect somatic copy number changes in low-depth sequencing data

## Overview

soma-cnv is a suite of tools to detect somatic copy number changes using low depth whole genome sequencing data.  Somatic copy number changes often occur in only a small proportion of the cells contributing to the sequenced DNA, and consequently manifest as subtle signals.  soma-cnv combines data from adjacent loci to increase its sensitivity to these subtle changes, trading positional resolution for sensitivity.  The output of soma-snv is a segmentation of the genome into copy number regions with estimated allele ploidies, and an estimation of the aneuploid fraction present in a sample.

Examples of somatic copy number changes detected by soma-cnv:

![Example of soma-cnv detections](/docs/example.png?raw=true "Somatic copy number changes detected by soma-cnv")


## Requirements

* A variant caller (eg GATK HaplotypeCaller, samtools.  Tested with GATK 3.7-0-gcfedb67)
* R (tested with v3.2.3)


## Workflow

The main soma-snv workflow (steps 2 and 3) operates on a per-sample basis.  If the generation of platform calibration files is required (step 1), this requires the analysis of many platform-matched samples in parallel.

1. (optional) Prepare platform-specific calibration files.
2. For each sample, collect allele-specific depths at whitelist loci.
3. For each sample, for a soma-cnv model and identify aneuploid regions.

## Example

### 1. Cohort: prepare whitelist and calibration files

soma-cnv requires platform-specific calibration files for accurate CNV detection.  Whitelist and calibration files suitable for human data generated on the HiSeqX sequencer from TruSeq Nano libraries, and mapped to hs37d5 with PhiX decoy, are supplied in data/.  For a different species, platform, or pipeline, new files are likely to be required.  This section describes the creation of these files.

Although soma-cnv runs in a single-sample mode, the generation of calibration files requires sequencing data from many (ideally over 100) platform-matched samples.  These samples must be from different individuals; the use of technical replicates to generate calibration data will reduce sensitivity.


#### A. Locus whitelist
The first file that needs to be created is a set of whitelist loci to use for CNV determination.  These loci should be a set of positions on the genome which can be genotyped with very high reliability on the chosen platform.  The details of whitelist definition will differ between platforms, but in general whitelist loci should have the following properties:
* Consistently high genotyping rate across multiple samples.
* Consistent and typical depth of sequencing across multiple samples.
* Autosomal, uniquely mapping, high mdust complexity, not in a repeat region.
* Intermediate GC content, for example in [0.3, 0.55] in a 100 bp window around the locus.
* Biallelic
* High variant allele frequency (>= 5%) in the cohort, and no excess heterozygosity (eg judged by HWE test)

Once whitelist variants have been identified, they should be saved to a tab-delimited table with header, in the following format: ```chrom  pos  ref  alt```.  Example:
```
chrom	pos	ref	alt
1	100000012	G	T
1	100000827	C	T
1	100002882	T	G
1	100004726	G	A
1	100005477	G	A
1	100006117	G	A
1	100006734	C	T
```

#### B. GC content
After loci have been defined, GC content covariates need to be calculated at these loci for per-sample GC correction.  The script in util/generate_gc.py has been written for this purpose.  Run as follows:
```
python generate_gc.py reference.fa whitelist.loci > whitelist.gc.tsv
```

Where ```reference.fa``` is the reference fasta, ```whitelist.loci``` is the set of whitelist loci from step A, and ```whitelist.gc``` is the gc covariate file.

#### C. Affinity correction


### 2. Per-sample: collection of allele-specific depths

This step examines the input mapped BAM file at the loci given in ```whitelist.loci```, and reports the total and alt allele depths at each heterozygous locus in the following tab-delimited table format:
```
<chrom>	<pos>	<dp>	<ad>
```
Note that this file has no header, and that homozygous reference or alternate loci are not included.  Example:
TODO

This file can be generated by a number of methods, and soma-cnv should be quite robust to the choice.  Example code for GATK HaplotypeCaller:
```
# Convert the tsv-format whitelist loci into an interval_list for use by GATK.
awk '{print $1 ":" $2 "-" $2}' < data/truseq_nano_hiseqX_hs37d5x.loci.tsv \
  > truseq_nano_hiseqX_hs37d5x.loci.interval_list

# Run GATK HC
# -ip 100 instructs HC to consider a region of 100 bp around each locus, to enable 
# local haplotype reassembly.  Note that because of this, some additional variant 
# loci may be reported (not just those in the interval_list), but these will be 
# removed at the later R stage.
java -Xmx2G -jar GenomeAnalysisTK.jar -T HaplotypeCaller -R <reference.fa> \
  -L truseq_nano_hiseqX_hs37d5x.loci.interval_list -ip 100 \
  -I sampleID.bam -o sampleID.temp.vcf

# Post-process the GATK VCF: extract het SNP loci and report depths.
awk -f util/filter_hc_vcf.awk < sampleID.temp.vcf \
  | xz -c \
  > sampleID.soma-cnv.hetdp.xz
```


### 3. Per-sample: Fit soma-cnv model and identify aneuploid regions

```
Rscript soma-cnv.R \
  data/truseq_nano_hiseqX_hs37d5x.affinity.tsv \
  data/truseq_nano_hiseqX_hs37d5x.gc.tsv \
  sampleID.soma-cnv.hetdp.xz sampleID.soma-cnv.rds
```

## Future
* soma-cnv currently has a specific failure mode in which germ-line amplifications are mis-called as somatic CNV.  Future work will extend the error model to specifically include this possibility and reduce the incidence of germline-driven false positives.

